python/xorbits/_mars/dataframe/base/applymap.py

```diff
@@ -0,0 +1,310 @@
+                return ret_dtypes, ret_index_value
+
+            output_type = new_dtypes = index_value = None
+
+        try:
+            empty_df = build_df(df, size=2)
+            with np.errstate(all="ignore"), quiet_stdio():
+                infer_df = empty_df.applymap(
+                    func,
+                    na_action=self.na_action,
+                    **self.kwds,
+                )
+            if index_value is None:
+                if infer_df.index is empty_df.index:
+                    index_value = "inherit"
+                else:  # pragma: no cover
+                    index_value = parse_index(pd.RangeIndex(-1))
+
+            output_type = output_type or OutputType.dataframe
+            new_dtypes = new_dtypes or infer_df.dtypes
+        except:  # noqa: E722  # nosec
+            pass
+
+        self.output_types = (
+            [output_type] if not self.output_types else self.output_types
+        )
+        dtypes = new_dtypes if dtypes is None else dtypes
+        index_value = index_value if index is None else parse_index(index)
+        return dtypes, index_value
+
+    def _call_dataframe(self, df, dtypes=None, dtype=None, name=None, index=None):
+        # for backward compatibility
+        dtype = dtype if dtype is not None else dtypes
+        dtypes, index_value = self._infer_df_func_returns(
+            df, dtypes, dtype=dtype, name=name, index=index
+        )
+
+        if index_value is None:
+            index_value = parse_index(None, (df.key, df.index_value.key))
+        for arg, desc in zip((self.output_types, dtypes), ("output_types", "dtypes")):
+            if arg is None:
+                raise TypeError(
+                    f"Cannot determine {desc} by calculating with enumerate data, "
+                    "please specify it as arguments"
+                )
+
+        if index_value == "inherit":
+            index_value = df.index_value
+
+        shape = df.shape
+
+        self.func = cloudpickle.dumps(self.func)
+
+        return self.new_dataframe(
+            [df],
+            shape=shape,
+            dtypes=dtypes,
+            index_value=index_value,
+            columns_value=parse_index(dtypes.index, store_data=True),
+        )
+
+    def __call__(
+        self, df, dtypes=None, dtype=None, name=None, index=None, skip_infer=False
+    ):
+        dtypes = make_dtypes(dtypes)
+        dtype = make_dtype(dtype)
+        self.func_token = get_func_token(self.func)
+        if skip_infer:
+            self.func = cloudpickle.dumps(self.func)
+            return self.new_dataframe([df])
+
+        return self._call_dataframe(
+            df, dtypes=dtypes, dtype=dtype, name=name, index=index
+        )
+
+
+def df_applymap(
+    df,
+    func,
+    na_action=None,
+    args=(),
+    dtypes=None,
+    dtype=None,
+    name=None,
+    output_type=None,
+    index=None,
+    skip_infer=False,
+    **kwds,
```

Comment:
Add type annoations for these parameters. And check other places when you define a func.